packages/apollo-engine-reporting/src/agent.ts

```diff
@@ -270,27 +325,42 @@ const serviceHeaderDefaults = {
 export class EngineReportingAgent<TContext = any> {
   private readonly options: EngineReportingOptions<TContext>;
   private readonly apiKey: string;
-  private logger: Logger = console;
-  private graphVariant: string;
-  private reports: { [schemaHash: string]: Report } = Object.create(
+  private readonly logger: Logger = console;
+  private readonly graphVariant: string;
+
+  private reports: { [executableSchemaId: string]: Report } = Object.create(
```

Comment:
(note: i wrote all the below before I realized that the only thing you did to these objects is prettify them. oops. well, it's still true.  and i still want to fix it but i don't want to ask you to, so, um, have https://github.com/apollographql/apollo-server/pull/4139)





If I understand correctly, there's a strong requirement here that reports, reportSizes, and reportHeaders all have the same keys all the time, and there's a bunch of implicit dependencies on that fact.

Can you instead combine them into a single map whose value type is a tiny class with three fields?

Additionally, the type `{[x: string]: Foo}` is actually problematic. It tells TypeScript "no matter what string I index by, I'll get a Foo out". But that's not how this data structure actually works! All of these should be changed into `{[x: string]: Foo | undefined}`, which in fact does reveal a bunch of places where we fetch an object from them and don't check to see if it was actually there...